### PR TITLE
controller: Add eviction support for detached volume

### PR DESCRIPTION
**Pull request by boknowswiki**
This change adds auto attach for eviction requested volume and auto
detach it once the eviction is done.

https://github.com/longhorn/longhorn/issues/1704

Signed-off-by: Bo Tao <bo.tao@rancher.com>

### CODE REVIEW

File: controller/volume_controller.go [hunk 1]
```diff
@@ -821,10 +825,13 @@ func (vc *VolumeController) ReconcileVolumeState(v *longhorn.Volume, es map[stri
 		return err
 	}
 
-	if err := vc.checkForAutoAttachment(v, e, rs, allScheduled); err != nil {
+	hasEvictionRequestedReplicas := vc.hasReplicaEvictionRequested(rs)
+
+	if err := vc.checkForAutoAttachment(v, e, rs, allScheduled, hasEvictionRequestedReplicas); err != nil {
```
Comment (yasker) [inline controller/volume_controller.go:830]: Both auto-attach and detach function can have replicas as a parameter, so we don't need to check the condition outside then pass it in.

File: controller/volume_controller.go [hunk 2]
```diff
@@ -1636,8 +1645,9 @@ func (vc *VolumeController) checkForAutoAttachment(v *longhorn.Volume, e *longho
 	// Do auto attachment for:
 	//   1. restoring/DR volumes.
 	//   2. offline expansion.
+	//	 3. Eviction requested on this volume.
```
Comment (yasker) [inline controller/volume_controller.go:1648]: Wrong indent.

File: controller/volume_controller.go [hunk 3]
```diff
@@ -1667,6 +1683,7 @@ func (vc *VolumeController) checkForAutoDetachment(v *longhorn.Volume, e *longho
 		v.Status.CurrentNodeID = ""
 		return nil
 	}
+
```
Comment (yasker) [inline controller/volume_controller.go:1686]: ?

**Review by yasker — CHANGES_REQUESTED** [review]

File: controller/volume_controller.go (hunk 1 above)
Comment (boknowswiki) [reply controller/volume_controller.go:830]: then the new function will be called twice. Each once in auto-attach and auto-detach, is that ok?

File: controller/volume_controller.go (hunk 2 above)
Comment (boknowswiki) [reply controller/volume_controller.go:1648]: I also noticed this, but in my local environment, the numbers are aligned. Let me double check on this.

File: controller/volume_controller.go (hunk 3 above)
Comment (boknowswiki) [reply controller/volume_controller.go:1686]: will remove this new line. Thanks.

File: controller/volume_controller.go (hunk 2 above)
Comment (yasker) [reply controller/volume_controller.go:1648]: It's normally tab vs space.

File: controller/volume_controller.go (hunk 1 above)
Comment (yasker) [reply controller/volume_controller.go:830]: That's fine since it's just a status check, doesn't have side effects.

Comment (boknowswiki) [reply controller/volume_controller.go:830]: Ok, will fix it.

Comment (boknowswiki) [reply controller/volume_controller.go:830]: Sorry my bad. auto-detach doesn't have replicas as parameter passed in. Should I add the rs new parameter or keep this way? Thanks.

Comment (yasker) [reply controller/volume_controller.go:830]: We can just add `rs` parameter. It's still a small change.

File: controller/volume_controller.go [hunk 4]
```diff
@@ -1617,7 +1621,8 @@ func (vc *VolumeController) reconcileVolumeSize(v *longhorn.Volume, e *longhorn.
 	return nil
 }
 
-func (vc *VolumeController) checkForAutoAttachment(v *longhorn.Volume, e *longhorn.Engine, rs map[string]*longhorn.Replica, allScheduled bool) error {
+func (vc *VolumeController) checkForAutoAttachment(v *longhorn.Volume,
```
Comment (yasker) [inline controller/volume_controller.go:1624]: Line breaks? Not sure if it's necessary.

Same below.

**Review by yasker — COMMENTED** [review]
Small coding style issue, otherwise LGTM.

@shuo-wu ?

File: controller/volume_controller.go (hunk 4 above)
Comment (boknowswiki) [reply controller/volume_controller.go]: Sure, just changed them back. Thanks.

**Review by yasker — DISMISSED** [review]

File: controller/volume_controller.go [hunk 5]
```diff
@@ -1657,6 +1662,10 @@ func (vc *VolumeController) checkForAutoDetachment(v *longhorn.Volume, e *longho
 
 	// Do auto-detachment for non-restore/DR volumes.
 	if !v.Status.RestoreRequired && !v.Status.IsStandby {
+		// Don't for auto-detachment if the evivtion is going on.
```
Comment (shuo-wu) [inline controller/volume_controller.go:1665]: Put it outside of this `if`. e.g.:
```
	if v.Status.ExpansionRequired {
		return nil
	}
       
        if vc.hasReplicaEvictionRequested(rs) {
		return nil
	}		
```

And there is a typo `evivtion` in the comment.

**Review by shuo-wu — CHANGES_REQUESTED** [review]

File: controller/volume_controller.go (hunk 5 above)
Comment (boknowswiki) [reply controller/volume_controller.go]: Fixed, thanks.

**Comment by yasker** [comment]
Since @shuo-wu 's comment has been addressed, I assume it's OK to merge this PR.

**Review by yasker — APPROVED** [review]

**Pull request merged**